d3a_interface/sim_results/kpi.py

```diff
@@ -162,45 +225,62 @@ def __repr__(self):
+                self.state[area_dict['uuid']].demanded_buffer_wh)
 
         # in case when the area doesn't have any load demand
         if self.state[area_dict['uuid']].total_demand <= 0:
             self_sufficiency = None
-        elif self.state[area_dict['uuid']].total_self_consumption_wh >= \
-                self.state[area_dict['uuid']].total_demand:
+        elif (self.state[area_dict['uuid']].total_self_consumption_wh >=
+              self.state[area_dict['uuid']].total_demand):
             self_sufficiency = 1.0
         else:
-            self_sufficiency = self.state[area_dict['uuid']].total_self_consumption_wh / \
-                               self.state[area_dict['uuid']].total_demand
+            self_sufficiency = (self.state[area_dict['uuid']].total_self_consumption_wh /
+                                self.state[area_dict['uuid']].total_demand)
 
         if self.state[area_dict['uuid']].total_energy_produced_wh <= 0:
             self_consumption = None
-        elif self.state[area_dict['uuid']].total_self_consumption_wh >= \
-                self.state[area_dict['uuid']].total_energy_produced_wh:
+        elif (self.state[area_dict['uuid']].total_self_consumption_wh >=
+              self.state[area_dict['uuid']].total_energy_produced_wh):
             self_consumption = 1.0
         else:
-            self_consumption = self.state[area_dict['uuid']].total_self_consumption_wh / \
-                               self.state[area_dict['uuid']].total_energy_produced_wh
+            self_consumption = (self.state[area_dict['uuid']].total_self_consumption_wh /
+                                self.state[area_dict['uuid']].total_energy_produced_wh)
         return {
             "self_sufficiency": self_sufficiency, "self_consumption": self_consumption,
             "total_energy_demanded_wh": self.state[area_dict['uuid']].total_demand,
             "demanded_buffer_wh": self.state[area_dict['uuid']].demanded_buffer_wh,
             "self_consumption_buffer_wh": self.state[area_dict['uuid']].self_consumption_buffer_wh,
             "total_energy_produced_wh": self.state[area_dict['uuid']].total_energy_produced_wh,
             "total_self_consumption_wh":
-                self.state[area_dict['uuid']].total_self_consumption_wh}
+                self.state[area_dict['uuid']].total_self_consumption_wh,
+            "base_case_revenue": getattr(self.saving_state.get(area_dict['uuid'], None),
+                                         'base_case_revenue', None),
+            "d3a_revenue": getattr(self.saving_state.get(area_dict['uuid'], None),
+                                   'd3a_revenue', None),
+            "saving_absolute": getattr(self.saving_state.get(area_dict['uuid'], None),
+                                       'saving_absolute', None),
+            "saving_percentage": getattr(self.saving_state.get(area_dict['uuid'], None),
+                                         'saving_percentage', None)
```

Comment:
I'm not sure I understand this part. Why are you using `getattr`? We know that all instances of `SavingsKPI` will always have these variables, since you've defined them in the `__init__()`: https://github.com/gridsingularity/d3a-interface/blob/2f2e624df987b329ece69544d13539f0712c452e/d3a_interface/sim_results/kpi.py#L26-L36.

What do you think about something like the following:
```suggestion
area_savings_kpi = self.saving_state.get(area_dict['uuid'])
return {
    ...
    "base_case_revenue": getattr(area_savings_kpi, 'base_case_revenue', None),
    "d3a_revenue": getattr(area_savings_kpi, 'd3a_revenue', None),
    "saving_absolute": getattr(area_savings_kpi, 'saving_absolute', None),
    "saving_percentage": getattr(area_savings_kpi, 'saving_percentage', None)
}
```
What do you think?

Reply:
Another thing that could be nice. You could add a `to_dict()` method in the `SavingsKPI` class, so that you would just have this:
```python
area_savings_kpi = self.saving_state.get(area_dict['uuid'])
area_savings_output = area_savings_kpi.to_dict()
return {
	    ...
	}.update(area_savings_output)
```

Reply:
> I'm not sure I understand this part. Why are you using `getattr`?
Purpose to use `getattr` is `saving_kpi` is only instantiated for house areas. Therefore it will raise the following exception for non-house areas.
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'abc'
```

Reply:
> Purpose to use getattr is saving_kpi is only instantiated for house areas. Therefore it will raise the following exception for non-house areas.

Thanks for the explanation :) I think this should be made explicit, and the two cases (house vs non-house areas) should be separated.
What is happening at the moment is basically this, for cases in which `SavingsKPI` hasn't been instantiated:
```python
getattr(None, "some_attribute", None)
```
This code is in my opinion too "broad", because it doesn't say much about the objects we're dealing with and it might lead to unexpected effects in the future. In theory, one could replace the values if `self.saving_state` with any possible object, and the code wouldn't raise any useful exception (making it impossible to debug).

If the area we're dealing with is not supposed to have a `savings_kpi` attribute (e.g. it's a house area), we should not try to query that attribute. For the same reason, the `KPI` instance of that area should not have a `savings_kpi` attribute, since the area doesn't support it.
One option would be to create separate classes (respecting the single-responsibility principle): one that deals with KPIs of house areas and another one that deals with other area types.

Reply:
Update: as you said, the `KPI` instance contains kpis for all the areas (not just one) so my suggestion of using two separate KPI classes doesn't probably apply. What I see is that the `area_performance_indices` method is doing "too much", so I have the feeling that it could be refactored. In any case, what I'm suggesting is to create more specialized code based on the type of area we're dealing with, and avoid the generic `getattr`.

Reply:
```
        kpi_parm_dict = {
            "self_sufficiency": self_sufficiency, "self_consumption": self_consumption,
            "total_energy_demanded_wh": self.state[area_dict["uuid"]].total_demand,
            "demanded_buffer_wh": self.state[area_dict["uuid"]].demanded_buffer_wh,
            "self_consumption_buffer_wh": self.state[area_dict["uuid"]].self_consumption_buffer_wh,
            "total_energy_produced_wh": self.state[area_dict["uuid"]].total_energy_produced_wh,
            "total_self_consumption_wh": self.state[area_dict["uuid"]].total_self_consumption_wh,
        }
        if self.savings_state.get(area_dict["uuid"]) is not None:
            kpi_parm_dict.update(self.savings_state[area_dict["uuid"]].to_dict())
        return kpi_parm_dict
```

@fievelk How about this for now?

Reply:
So if `savings_state[area_dict["uuid"]]` is not there, it means that the area is a house (and it doesn't have KPIs for savings). Am I correct? If so, adding a comment to explain this would be great. In any case, this looks like a good way to go in my opinion 👍 Thanks for dealing with it!

Reply:
> So if savings_state[area_dict["uuid"]] is not there, it means that the area is`n't` a house (and it doesn't have KPIs for savings).

Minor correction(hope that was a typo)

Reply:
Yes: `savings_kpi` is only instantiated for house areas, so if it's not there it means that the area is not a house. Thanks for the correction! 👍